src/parser/parser.py

```diff
@@ -0,0 +1,161 @@
+        self.errors: list[str]
+        self.pos: int = 0
+
+        self.parse_program()
+
+    @property
+    def curr_tok(self) -> Token:
+        'returns none if no more tokens'
+        if self.pos >= len(self.tokens):
+            return Token("", TokenType.EOF, (0, 0), (0, 0))
+        return self.tokens[self.pos]
+
+    def peek_tok(self) -> Token:
+        'returns none if no more tokens'
+        if self.pos + 1 >= len(self.tokens):
+            return Token("", TokenType.EOF, (0, 0), (0, 0))
+        return self.tokens[self.pos + 1]
+
+    def advance(self) -> int | None:
+        'returns none if can not advance'
+        if self.pos >= len(self.tokens):
+            return None
+        self.pos += 1
+
+    def parse_program(self) -> Program:
+        '''
+        parse program production
+
+        on a global scope, only function, classes, and variable/constant
+        declarations are allowed. expressions are not allowed
+        '''
+        p = Program()
+        while self.curr_tok.token != TokenType.EOF:
+            if self.curr_tok.token == TokenType.FWUNC:
+                p.fwuncs.append(self.parse_fwunc())
+
+            elif self.curr_tok.token == TokenType.CWASS:
+                p.cwass.append(self.parse_cwass())
+
+            elif self.curr_tok.token == TokenType.GWOBAW:
+                p.gwobaws.append(self.parse_gwobaw_dec())
+
+            else:
+                # TODO: add error messages for unexpected tokens in global scope
+                # up to you if you want to be super specific and provide hints
+                # based on the token type encountered
+                self.advance()
+
+        p.print()
+        return p
+
+    def parse_fwunc(self) -> Fwunc:
+        '''
+        parse fwunc production
+
+        function declaration example:
+            fwunc mainuwu-san(name-senpai) [[
+                pwint("hewwo |name|!")~
+            ]]
+
+        will early return if found EOF
+        '''
+        f = Fwunc()
+
+        self.advance()
+        if self.pos is None:
+            self.errors.append("expected function declaration, got EOF")
+            return f
+
+        if self.curr_tok.token.token.startswith("IDENTIFIER"):
+            f.id = self.curr_tok
+            self.advance()
+            if self.pos is None:
+                self.errors.append(f"expected function declaration for {f.id.lexeme}, got EOF")
+                return f
+        else:
+            self.errors.append(f"functions must have an identifier, got '{self.curr_tok.lexeme}'")
+
+        if self.curr_tok.token == TokenType.DASH:
+            self.advance()
+            if self.pos is None:
+                self.errors.append(f"expected type after '-' for function declaration, got EOF")
+                return f
+        else:
+            self.errors.append(f"functions must have a dash before a type, got '{self.curr_tok.lexeme}'")
+
+        data_types = [
```

Comment:
question, i-lilipat ba to sa productions?

Reply:
Sure but how do you suggest implementing it?

Reply:
right now, wala pa, I may need to explore more your initial implementation especially on productions since may bagong python syntax akong nakikita hahaha.

Reply:
Looks good naman so far, baka lang may mag bago bigla once ituro ni maam predict set bukas. I am on board naman w/ this approach. If I find something while exploring it, imma make suggestions